include/communicator_quda.h

```diff
@@ -0,0 +1,963 @@
+#pragma once
+
+#include <unistd.h> // for gethostname()
+#include <assert.h>
+#include <limits>
+
+#include <quda_internal.h>
+#include <comm_quda.h>
+#include <csignal>
+
+#include <algorithm>
+#include <numeric>
+
+#if defined(MPI_COMMS) || defined(QMP_COMMS)
+#include <mpi.h>
+#endif
+
+#if defined(QMP_COMMS)
+#include <qmp.h>
+#endif
+
+#ifdef QUDA_BACKWARDSCPP
+#include "backward.hpp"
+namespace backward
+{
+  static backward::SignalHandling sh;
+} // namespace backward
+#endif
+
+struct Topology_s {
+  int ndim;
+  int dims[QUDA_MAX_DIM];
+  int *ranks;
+  int (*coords)[QUDA_MAX_DIM];
+  int my_rank;
+  int my_coords[QUDA_MAX_DIM];
+  // It might be worth adding communicators to allow for efficient reductions:
+  //   #if defined(MPI_COMMS)
+  //     MPI_Comm comm;
+  //   #elif defined(QMP_COMMS)
+  //     QMP_communicator_t comm; // currently only supported by qmp-2.4.0-alpha
+  //   #endif
+};
+
+static const int max_displacement = 4;
+
+inline int lex_rank_from_coords_dim_t(const int *coords, void *fdata)
+{
+  int *dims = reinterpret_cast<int *>(fdata);
+  int rank = coords[0];
+  for (int i = 1; i < 4; i++) { rank = dims[i] * rank + coords[i]; }
+  return rank;
+}
+
+inline int lex_rank_from_coords_dim_x(const int *coords, void *fdata)
+{
+  int *dims = reinterpret_cast<int *>(fdata);
+  int rank = coords[3];
+  for (int i = 2; i >= 0; i--) { rank = dims[i] * rank + coords[i]; }
+  return rank;
+}
+
+/**
+ * Utility function for indexing into Topology::ranks[]
+ *
+ * @param ndim  Number of grid dimensions in the network topology
+ * @param dims  Array of grid dimensions
+ * @param x     Node coordinates
+ * @return      Linearized index cooresponding to the node coordinates
+ */
+static inline int index(int ndim, const int *dims, const int *x)
+{
+  int idx = x[0];
+  for (int i = 1; i < ndim; i++) { idx = dims[i] * idx + x[i]; }
+  return idx;
+}
+
+static inline bool advance_coords(int ndim, const int *dims, int *x)
+{
+  bool valid = false;
+  for (int i = ndim - 1; i >= 0; i--) {
+    if (x[i] < dims[i] - 1) {
+      x[i]++;
+      valid = true;
+      break;
+    } else {
+      x[i] = 0;
+    }
+  }
+  return valid;
+}
+
+// QudaCommsMap is declared in quda.h:
+//   typedef int (*QudaCommsMap)(const int *coords, void *fdata);
```

Comment:
???

Reply:
It's basically saying what `QudaCommsMap` is and where it's defined.